pinot-spi/src/main/java/org/apache/pinot/spi/accounting/ThreadResourceUsageProvider.java

```diff
@@ -0,0 +1,145 @@
+        && IS_THREAD_ALLOCATED_MEMORY_ENABLED;
+  }
+
+  public long getThreadTimeNs() {
+    return _isThreadCpuTimeMeasurementEnabled ? MX_BEAN.getCurrentThreadCpuTime() - _startTimeNs : 0;
+  }
+
+  public long getThreadAllocatedBytes() {
+    try {
+      return _isThreadMemoryMeasurementEnabled ? (long) SUN_THREAD_MXBEAN_GET_BYTES_ALLOCATED_METHOD
+          .invoke(MX_BEAN, Thread.currentThread().getId()) - _startBytesAllocated : 0;
+    } catch (IllegalAccessException | InvocationTargetException e) {
+      LOGGER.error("Exception happened during the invocation of getting initial bytes allocated", e);
+      return 0;
+    }
+  }
+
+  //initialize the com.sun.management.ThreadMXBean related variables using reflection
+  static {
+    Class<?> sunThreadMXBeanClass;
+    try {
+      sunThreadMXBeanClass = Class.forName(SUN_THREAD_MXBEAN_CLASS_NAME);
+    } catch (ClassNotFoundException e) {
+      LOGGER.error("Not able to load com.sun.management.ThreadMXBean, you are probably not using Hotspot jvm");
+      sunThreadMXBeanClass = null;
+    }
+
+    boolean isThreadAllocateMemorySupported = false;
+    try {
+      isThreadAllocateMemorySupported =
+          sunThreadMXBeanClass != null && (boolean) sunThreadMXBeanClass
+              .getMethod(SUN_THREAD_MXBEAN_IS_THREAD_ALLOCATED_MEMORY_SUPPORTED_NAME)
+              .invoke(MX_BEAN);
+    } catch (IllegalAccessException | NoSuchMethodException | InvocationTargetException e) {
+      LOGGER.error("Not able to call isThreadAllocatedMemorySupported, ", e);
+    }
+    IS_THREAD_ALLOCATED_MEMORY_SUPPORTED = isThreadAllocateMemorySupported;
+
+    boolean isThreadAllocateMemoryEnabled = false;
+    try {
+      isThreadAllocateMemoryEnabled =
+          sunThreadMXBeanClass != null && (boolean) sunThreadMXBeanClass
+              .getMethod(SUN_THREAD_MXBEAN_IS_THREAD_ALLOCATED_MEMORY_ENABLED_NAME)
+              .invoke(MX_BEAN);
+    } catch (IllegalAccessException | NoSuchMethodException | InvocationTargetException e) {
+      LOGGER.error("Not able to call isThreadAllocatedMemoryEnabled, ", e);
+    }
+    IS_THREAD_ALLOCATED_MEMORY_ENABLED = isThreadAllocateMemoryEnabled;
+
+    Method threadAllocateBytes = null;
+    if (IS_THREAD_ALLOCATED_MEMORY_SUPPORTED && IS_THREAD_ALLOCATED_MEMORY_ENABLED) {
+      try {
+        threadAllocateBytes = sunThreadMXBeanClass
+            .getMethod(SUN_THREAD_MXBEAN_GET_BYTES_ALLOCATED_NAME, long.class);
+      } catch (NoSuchMethodException ignored) {
+      }
+    }
+    SUN_THREAD_MXBEAN_GET_BYTES_ALLOCATED_METHOD = threadAllocateBytes;
+  }
+
+  static {
+    LOGGER.info("Current thread cpu time measurement supported: {}", IS_CURRENT_THREAD_CPU_TIME_SUPPORTED);
+    LOGGER.info("Current thread allocated bytes measurement supported: {}", IS_THREAD_ALLOCATED_MEMORY_SUPPORTED);
+    LOGGER.info("Current thread allocated bytes measurement enabled: {}", IS_THREAD_ALLOCATED_MEMORY_ENABLED);
+  }
```

Comment:
Curious why unlike THREAD_CPU_TIME, THREAD_ALLOCATED_MEMORY needs to have separate info on whether it is enabled and supported ?

Reply:
Refactored the code on this in the latest commit. `IS_THREAD_ALLOCATED_MEMORY_ENABLED` is a switch that can be turned on during runtime.